### ISSUE TITLE
Custom scrollbar

### ISSUE THREAD

**Issue by ericscheid** [naturalcrit/homebrewery#3355]
### Your idea:

[5e-Cleric](https://github.com/5e-Cleric) writes..

> PR #3346 adds in a scrollbar that i personally like more than the basic one:
> 
> <img width="284" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/91662974/59342880-7dd4-4503-854a-b5c14b645c39">
> 
> The color is the inverse color of the background. 
> 
> Everyone will have an opinion on the color, so feel free to comment.
> 
> A border radius can be added, but sacrificing the space between the bar and the edge of the screen

**Comment by ericscheid** [comment]
Adding as an issue for ease of finding in the future. 

(All PRs really should start from an issue, fwiw)

**Comment by calculuschild** [comment]
Making note of this exchange about having the scrollbar auto-hide, so it doesn't get lost:

> On MacOS, I have the option of having scroll bars auto-hide, but this css has the thumb always visible. Blech.

> It would likely need a onScroll event on the brewRenderer to change the opacity of the scroll bar (with a fade transition to make it smooth).

> Right. I don't think there is a CSS-only way to detect the scrolling state. It is possible with Javascript. Found this example that uses event listeners:
>
> ```
>    const scrollableDiv = document.getElementById("scrollable");
>    let scrollingTimeout;
>
>    scrollableDiv.addEventListener("scroll", () => {
>      clearTimeout(scrollingTimeout);
>      scrollableDiv.style.setProperty("--scrollbar-thumb-color", "black");
>      scrollingTimeout = setTimeout(() => {
>        scrollableDiv.style.setProperty("--scrollbar-thumb-color", "rgba(90, 90, 90, 0.5)");
>      }, 300); // Adjust the delay as needed
>    });
> ```

**Comment by G-Ambatte** [comment]
A minor layout issue with the new scrollbar - the page numbers appear under the scroll bar.

![image](https://github.com/naturalcrit/homebrewery/assets/19826659/8ebfbe72-9894-4e59-aaa3-2e533c3224ff)

**Comment by Gazook89** [comment]
I know this is merged already, and there was an opportunity to comment before it was, so this is sort of pissing in the wind--- but I'm not a huge fan of this change.  

1. The width used for a scrollbar is greater than the native scrollbar, *and* it overlaps the writing which is particularly bad in the editor, beyond just what @G-Ambatte notes above.  I think this could be improved by dropping the 'overlay' effect and just accepting the extra 'breathing room' that a hidden scroll track offers.  Remove this part:

```css
.CodeMirror-sizer {
    padding-right: 0 !important;
```

So instead of this:

<img width="359" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/379df1c1-5978-40eb-aafe-4bf1ecc8198a">

there is this:
<img width="359" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/479122d9-fef3-4246-831a-80349e54246c">


2. I do think there is something to be said for auto-hiding the scrollbar, but I also know that doing so can illicit some strong feelings from internet denizens, and is also less accessible.  

3. There are also some more technical issues with the current merged change.  It seems like in order to get a custom 'width' of the scroll bar thumb, the `background` property of the thumb uses a `linear-gradient` with a sharp cut to transparent at 15px (so the thumb is only 15px wide).  This works on vertical scrollbars, but doesn't work well with horizontal bars.  Here is an example of what the scrollbar looks like on the live site, followed by an example of what the same scrollbar looks like with the merged changes:

live:
<img width="359" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/6f938c8b-5bc4-4604-a8b6-5f7d7119943a">

with changes:
<img width="359" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/a8945721-1385-4b91-a379-19428389a70e">

The thumb bit only shows a 15 x 15 square, but it is *actually* the same width as the thumb in the original.  Which means that I can only move the visible part of the thumb a fraction of the width of the page.  This made me think it was broken until I took a few closer looks.  The furthest *right* I can move the thumb:

<img width="359" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/2bc4744b-033e-4a55-b1cd-9e740a7a62d9">

4. There is an annoying white square in the bottom right of the preview pane, where two scroll bars meet (if your window requires both a vertical and horizontal bar).  I don't know how to avoid that.  It's not present in the live site likely because the scrollbar tracks are white.

**Comment by 5e-Cleric** [comment]
All adressed in #3388

**Comment by 5e-Cleric** [comment]
For clarifying, i don't think we want to auto-hide the scrollbar at all, it is less accessible. But if you Really want it, i will implement it.

**Comment by Gazook89** [comment]
> auto-hiding the scrollbar

I agree, there isn't much value to this.  

> There is an annoying white square in the bottom right of the preview pane,

I see that it isn't there anymore, using `&-corner` and setting it to hidden.  It still looks odd to me in the way the default scrollbars don't.  Have a track color + a matching corner color feels better, to me.  Somehow, more solid, something that the thumb 'bumps' into.  I made an example below.  I added a border to the track, only to the top edge of the horizontal scroll and the left edge of the vertical scroll.  I also set the corner color to something slightly darker than the thumb.  And I reduced the width to 15px.  And I temporarily removed the '.pageInfo' div so it shows better:

<img width="233" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/54ff262c-5c39-4aab-9960-5ecbf3f9d4e4">

Speaking of `.pageInfo` div....this would be more work, but I think the solution that I would be most enthused about is to get the page info box into the bottom right corner, and become a sort of 'toolbar' in the future.  This would supplant the need for a "corner box", since it would become a permanent corner box.  The trick is that the scroll bars might need to be custom JS thing, since their width wouldn't be the true width of the scrolled element (in this case, the scroll bar wouldn't be the same width as the pane).

<img width="969" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/71aab0a3-0699-4914-9cd8-ce520fc5460f">


Moving on, I'm not sure if it's intentional to have this little bit of color change in the scroll thumb?  It's because the height/width of the scrollbar tracks are set to 20px, and the thumb gradient steps are at 15px.  I would suggest not using a gradient at all, and using a flat color since the rest of the UI just uses flat design.
<img width="34" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/e253d594-dbb5-4281-b9c4-55de749275e4">

**Comment by 5e-Cleric** [comment]
The reason for the gradient on itself is to visually separate the scrollbar from the window border, that is the only CSS way i have found to do so.

I love your idea for the js scrollbar to incorporate pageinfo, looks cool, but don't think i have the time to make it. Also that might just be our lowest priority.

About the corner, am i the only one who prefers scrollbars without tracks? when the track is obviously the page entire size, it looks very antiquated to me. If it were up to me, i would leave it as overlay without tracks.

**Comment by Gazook89** [comment]
I personally just like the normal scrollbars because they are so common and basically 'disappear' from my vision even if they don't actually disappear.  Anything that draws my attention to them is bad.  So my first preference is to leave them untouched.

I do like a track, as it provides a hard edge for the document (since I also don't like to have my scrollbar overlapping what i'm reading/writing).  

And I prefer a narrow scrollbar rather than wider, for screen real estate reasons-- but honestly it's likely the "what I am used to" factor too.

In short, I don't think any change to the scrollbars is necessary or desired.  But I also am just one person and don't want to squash dreams.  Mostly I'm just raising these points so that there is some record of possible objections, rather than just a "no one commented, so it must be good" process.

**Comment by Gazook89** [comment]
It seems like it might be possible to get a shorter scrollbar that only spans a fraction of the scrolled element, but I'm still researching it.  There are webkit-only solutions (like most of the scrollbar styling options), as seen in this [fiddle](https://jsfiddle.net/6rpzxuqg/1/), but it's not perfect and i hesitate to reach for chrome-only things when it comes to simple UI choices.

**Comment by 5e-Cleric** [comment]
It's a scrollbar, not a dream. We could revert to the original and forget about this headache.

**Comment by Gazook89** [comment]
Can the scrollbars be reverted, and such a change be left for users styles (via Stylish extension or similar)?

**Comment by 5e-Cleric** [comment]
> Can the scrollbars be reverted, and such a change be left for users styles (via Stylish extension or similar)?
> 
> Found another issue tonight: the CodeMirror lines don't stretch to the edge of the pane. I assume there is some hardcoded width for these lines that is somewhere in CM, but I'm not sure where.
> 
> <img alt="image" width="299" src="https://private-user-images.githubusercontent.com/58999374/326680690-c7091a6d-3bf6-4805-81e8-59cdd84d0cab.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ1OTMxOTgsIm5iZiI6MTcxNDU5Mjg5OCwicGF0aCI6Ii81ODk5OTM3NC8zMjY2ODA2OTAtYzcwOTFhNmQtM2JmNi00ODA1LTgxZTgtNTljZGQ4NGQwY2FiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAxVDE5NDgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjOGFhY2U3YWEyOGEwODQ3ZWVjOWUzMjRmMDY3MTk0ZDdhNDU0ZWMwN2M4YTk2ZDcwYzMxMWJlOGIyZGYxYmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DDQIPLntxzVe8FkxP939c1m0GVJe3VNUaWaA5Bil0mg">

John? you asked for this specifically a few weeks ago https://github.com/naturalcrit/homebrewery/issues/3355#issuecomment-2026655526

**Comment by Gazook89** [comment]
Yeah I suppose in a way I did, but I don't think this is exactly a reversal of my earlier position, just an addendum.

A solid spanning decoration like the page break line highlight seems odd if it stops just short of the **"visual edge"** of the editor.   I know earlier that I didn't want to have a scrollbar overlapping my text content but I think this is something different than "my content".  Covering my written text, which I may be actively reading or even editing, is visually intrusive.  Covering an editor decoration, on the other hand, doesn't seem as big of a problem.  Regarding covering the page number text...eh, not in love with it, but I still think stretching the page line break to the "visual edge" is higher priority.

So my preference would be:
- no scrollbar covering my written text
- no strange gaps in full width line highlights and the edge of the editor

Native scrollbars achieve both-- nothing covers content, and the visible track edges provide a different "visual edge" to the editor to which the full width line highlights can span and connect to.  

Native scrollbars are obviously my preference, but in the interest of being open to different options, maybe the below will help with the current implementation:

Change the line highlight for the page breaks to have a width of 110% or similar.

<img width="528" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/e010ab35-925f-46bf-8fe8-4e9d83a98354">

It'd be here:   https://github.com/NaturalCrit/homebrewery/blob/master/client/homebrew/editor/editor.less#L8
